docs: initial rover dev (#1323)
apollographql/apollo-cli: docs/source/commands/dev.mdx

---
title: The Rover dev command

> description: Query multiple GraphQL APIs from localhost

Reviewer: Okay, this might be a bit wacky, but what do you think about something like this:
```suggestion
description: Query a locally federated GraphQL API
```

Author: looks good!

---

> 

Reviewer: I'd recommend adding a quick visual reminder of what subgraphs and supergraphs are (like we do in ["Rover subgraph commands"](https://www.apollographql.com/docs/rover/commands/subgraphs)).

> `rover dev` starts a local router that can query across one or more running GraphQL APIs (subgraphs)

Reviewer: ```suggestion
You can use Rover to start a local router that can query across one or more running GraphQL APIs (i.e., subgraphs) through a single endpoint (i.e., supergraph). 
```

> This Rover command allows you to federate one or more running subgraphs into a supergraph that can be queried by any GraphQL client.

Reviewer: I took a stab at breaking this out into two sentences; feel free to tweak if this isn't technically accurate 😅 : 
```suggestion
You can use Rover to federate one or more running subgraphs into a supergraph with the `dev` command.  Once Rover spins up a local Apollo Router instance and composes your new supergraph, any GraphQL client can query your new local endpoint. 
```

Reviewer: It might be nice to include a quick recap above this to remind folks of what a subgraph and a supergraph are, like the ["Rover supergraph commands" article](https://www.apollographql.com/docs/rover/commands/supergraphs) does.

> 

Reviewer: I'd suggest adding a second-level header here:
```suggestion

## Creating a federated session

```

Do we have an overall term for a rover `dev` setup (with the Router + supergraph)? Do we want to call this a "session"?

Author: federated session seems good to me

> through one endpoint (supergraph). As you add, edit, and remove subgraphs, `rover dev` automatically

Reviewer: ```suggestion
As you add, edit, and remove subgraphs, Rover automatically composes their separate schemas into a unified supergraph schema that the router reloads and serves. 
```

> The first `rover dev` process you run starts a dev instance of the Apollo Router and connects it to

Reviewer: (Note: I had to trim off one of the backticks from the below `bash` command for this suggestion block to work. Sorry!)

Took a stab at splitting up paragraph a bit, let me know what you think: 
```suggestion
Rover's `dev` command intakes a running subgraph, for example:

```bash
rover dev --name exampleSubgraph --schema ./schema.graphql --url http://localhost:4000
``

The first time you run the `dev` command, Rover starts a process that: creates a development instance of the [Apollo Router](/router), makes a supergraph based on the subgraph you specified, and connects the new router to the new supergraph.

You can add subgraphs to your local supergraph by running additional processes using the `dev` command. As you add subgraphs, Rover automatically composes each subgraph schema into a new supergraph schema, all hosted by the *same router instance*.
```

Think plug-n-play USB devices but with your GraphQL APIs!

> ⚠️ Do not run this command in production! ⚠️ It is intended for local development.

Reviewer: ```suggestion
> ⚠️ Do not run `rover dev` in production! ⚠️ This command is only intended for local development.
```

> The router instance is tied to the *first* `rover dev` process. If you terminate that process, the

Reviewer: Suggested style tweaks for this paragraph: 
```suggestion
Your router instance is tied to the *first* process running `rover dev`. Terminating the initial process closes the session, ending the router and any attached processes. If you terminate a `rover dev` process *besides* the first process (i.e., removing a subgraph), Rover composes a new supergraph schema from the remaining subgraphs.
```

router terminates along with any attached `rover dev` processes. If you terminate a `rover dev` 
process *besides* the first process (thereby removing a subgraph), a new supergraph schema is 
composed from the remaining subgraphs.

> ## Starting the first `rover dev` session

Reviewer: ```suggestion
### Starting the session
```

> Before starting `rover dev`, you need to start your subgraph server. Note the URL it is running on, and the location of the schema on the filesystem if it exists. If you don't have a local schema file, introspection will need to be enabled on the subgraph.

Reviewer: ```suggestion
To use Rover's dev' command, you'll first need access to a running subgraph server. You'll need the URL endpoint the subgraph is serving on and the location of the schema file in the filesystem. If you don't have a local schema file, you can instead [enable introspection](/apollo-server/api/apollo-server#introspection) for your subgraph.
```

> An example `rover dev` invocation might look something like this:

Reviewer: ```suggestion
An example `rover dev` command for running a local subgraph might look something like this:
```

`npm run start && rover dev --name products --schema ./products.graphql --url http://localhost:4000`

> This will start the main `rover dev` session, watching `./products.graphql` for changes to the schema, and start up a local Apollo Router that can be queried at http://localhost:3000. This session will start listening for other `rover dev` sessions that include other subgraph schemas.

Reviewer: ```suggestion
The above command starts the `rover dev` session, watching `./products.graphql` for changes to the schema, and starting up a local Apollo Router at the default endpoint (i.e.,`http://localhost:3000`). This session also listens for other `rover dev` sessions that include different subgraph schemas.
```

> ### Kill the first `rover dev` session

Reviewer: ```suggestion
### Ending the session
```

> If you kill the first `rover dev` session, it will shut down the router and any attached `rover dev` sessions will be detached and shut down.

Reviewer: ```suggestion
If you terminate the initial `rover dev` process, it shuts down the router, and any attached `rover dev` processes are detached and shut down.
```

> ## Attaching more subgraphs to the main `rover dev` session

Reviewer: ```suggestion
### Attaching additional subgraphs to a session
```

> For each new subgraph, run `rover dev` in a new terminal with the necessary arguments that refer to your GraphQL API (subgraph). If the `--supergraph-port` argument is equivalent to a running `rover dev` session, the new session will attach to the running session and provide it with updates to the subgraph schema. The main session will then re-compose the supergraph schema, printing the build result in the output of the main session and the attached session.

Reviewer: Can we take out the `--supergraph-port` argument part here? Maybe move it to a new section? 

Also, how does this style tweak feel:
```suggestion
To add a new subgraph to a session, run `rover dev` in a new terminal with the necessary arguments for your GraphQL API (i.e., subgraph). If a session is already running, Rover automatically attaches subsequent `rover dev` processes to that existing session. 

When you add a new subgraph to a session, Rover handles recomposing the supergraph schema and printing the build result in the outputs of both the main session and the attached session.
```

> ## Triggering composition on subgraph changes

Reviewer: ```suggestion
### Triggering recomposition on subgraph changes
```

> Each `rover dev` session is responsible for monitoring its subgraph schema.

Reviewer: ```suggestion
Each `rover dev` process is responsible for monitoring its subgraph schema.
```

> ### Via filesystem update _(recommended)_

Reviewer: ```suggestion
#### Via filesystem update _(recommended)_
```

> If you pass the `--schema <PATH>` argument, `rover dev` will watch the schema file for changes and report the schema update to the main `rover dev` session when its contents change.

Reviewer: ```suggestion
If you pass the `--schema <PATH>` argument, `rover dev` will watch the schema file for changes and report any updates to the main `rover dev` session if a schema's contents change.
```

> ### Via introspection _(not recommended)_

Reviewer: ```suggestion
#### Via introspection _(not recommended)_
```

> If you do not pass the `--schema <PATH>` argument, `rover dev` will introspect the URL that was specified by the `--url <SUBGRAPH_URL>` argument once every second. The command prefers to use the same query as `rover subgraph introspect`, but it will fall back to `rover graph introspect` if your GraphQL API does not support introspecting federated SDL. Note that in this case directives will be stripped from the subgraph schema.

Reviewer: ```suggestion
If you don't pass the `--schema <PATH>` argument, `rover dev` will introspect the URL specified by the `--url <SUBGRAPH_URL>` argument once every second. The command prefers to use the same query as `rover subgraph introspect`, but it will fall back to `rover graph introspect` if your GraphQL API does not support introspecting federated SDL. If this is the case, directives are stripped from the subgraph schema.
```

> ### Stopping an attached `rover dev` session

Reviewer: ```suggestion
### Stopping an attached process
```

> If you stop an attached `rover dev` session (by pressing ctrl+c), the supergraph will de-compose that subgraph and reload the router.

Reviewer: ```suggestion
If you stop an attached `rover dev` process (using `ctrl+c` or closing the terminal), the supergraph will de-compose the removed subgraph and reload the router.
```

Author: closing the terminal abruptly actually doesn't allow the "remove this subgraph" message to actually be sent. i may need to do some thinking on that one.

---
title: The Rover dev command
description: Query a locally federated GraphQL API
---

You can use Rover to start a local router that can query across one or more running GraphQL APIs (i.e., subgraphs)
through one endpoint (i.e., supergraph). As you add, edit, and remove subgraphs, Rover automatically
composes their separate schemas into a unified supergraph schema that the router reloads and serves. 

Think plug-n-play USB devices but with your GraphQL APIs!

> A **subgraph** is a graph that contributes to the composition of a federated **supergraph**:

Reviewer: How would you feel about moving this section above the first paragraph? So folks have the context of subgraph v. supergraph, then get introduced to how Rover dev comes into play

```mermaid
graph BT;
  gateway(["Supergraph (A + B + C)"]);
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  serviceC[Subgraph C];
  gateway --- serviceA & serviceB & serviceC;
```

> > ⚠️ Do not run this command in production! ⚠️ This command is only intended for local development.

Reviewer: On second thought, maybe we should move this warning to the top of the article. So that folks know right from the start that Rover dev is for local dev only. 

Also a slight style tweak:

```suggestion
> ⚠️  **Note:** The `rover dev` command is _only_ for local development. Do **not** run this command in production.
```

## Creating a federated session

### Starting the session

> Before starting `rover dev`, you need to start your GraphQL API (subgraph). Note the URL it is running on, and the location of the schema on the filesystem if it exists. If you don't have a local schema file, introspection will need to be enabled on the subgraph.

Reviewer: Took a second pass for style:
```suggestion
Before using Rover's `dev` command, you first need to have access to a running GraphQL API (subgraph). You'll need the URL the GraphQL API is serving and the location of the schema in the filesystem. If you don't have a local schema file, your subgraph must enable introspection.
```

An example `rover dev` command for running a local subgraph might look something like this:

`rover dev --name products --schema ./products.graphql --url http://localhost:4000`

> 

Reviewer: Should we explicitly clarify somewhere if the GraphQL APIs _themselves_ have to be local? As in, can you use `rover dev` with a hosted GraphQL API running somewhere else (in which case I guess you'd have to know what introspection was on for that subgraph)?

The above command starts the `rover dev` session, watching `./products.graphql` for changes to the schema, and starts up a local Apollo Router at the default endpoint (i.e., `http://localhost:3000`). This session also listens for other `rover dev` processes that include different subgraph schemas.

### Stopping the session

> If you stop the first `rover dev` session (by pressing ctrl+c), it will shut down the router and any attached `rover dev` sessions will be detached and shut down.

Reviewer: ```suggestion
If you stop the first `rover dev` session (pressing `CTRL+C`), it will shut down the router, and any attached `rover dev` processes are detached and shut down.
```

> If you stop the first `rover dev` session (by pressing `CTRL+C`), it will shut down the router, and any attached `rover dev` processes are detached and shut down.

Reviewer: Sorry to keep adding stuff! How do you feel about this slight tweak:
```suggestion
If you stop the first `rover dev` process (by pressing `CTRL+C`), it will shut down the router, and any attached `rover dev` processes are detached and shut down as the session ends.
```

### Attaching additional subgraphs to a session

> To add a new subgraph to a session, run `rover dev` in a new terminal with the necessary arguments for your GraphQL API (i.e., subgraph). If a session is already running, Rover automatically attaches subsequent `rover dev` processes to that existing session.

Reviewer: ```suggestion
To add a new GraphQL API (i.e., subgraph) to a session, run `rover dev` in a new terminal with the necessary arguments for that new subgraph. If a session is already running, Rover automatically attaches subsequent `rover dev` processes to that existing session.
```

> When you add a new subgraph to a session, Rover handles recomposing the supergraph schema and printing the build result in the outputs of both the main session and the attached session.

Reviewer: Changing these to process(along the lines of each `rover dev` process is part of the unified session)
```suggestion
When you add a new subgraph to a session, Rover handles recomposing the supergraph schema and printing the build result in the outputs of both the primary process and the attached process.
```

### Triggering recomposition on subgraph changes

> Each `rover dev` process is responsible for monitoring its subgraph schema.

Reviewer: How would you feel about making this a bit more explicit:
```suggestion
Each `rover dev` process is responsible for monitoring any changes to its subgraph schema.
```

#### Via filesystem update _(recommended)_

> If you pass the `--schema <PATH>` argument, `rover dev` will watch the schema file for changes and report any updates to the main `rover dev` session if a schema's contents change.

Reviewer: ```suggestion
If you pass the `--schema <PATH>` argument, `rover dev` watches that schema file for changes and reports any updates to the main `rover dev` process if a schema's contents change.
```

> #### Via introspection _(not recommended)_

Reviewer: I think we can remove the explicit "not recommended" because there are only two options 😂 :
```suggestion
#### Via introspection 
```

> If you don't pass the `--schema <PATH>` argument, `rover dev` will introspect the URL that was specified by the `--url <SUBGRAPH_URL>` argument once every second. The command prefers to use the same query as `rover subgraph introspect`, but it will fall back to `rover graph introspect` if your GraphQL API does not support introspecting federated SDL. If this is the case, directives are stripped from the subgraph schema.

Reviewer: ```suggestion
If you don't pass the `--schema <PATH>` argument, `rover dev` will introspect the URL specified by the `--url <SUBGRAPH_URL>` argument once every second. The command prefers to use the same query as `rover subgraph introspect` but falls back to `rover graph introspect` if your GraphQL API doesn't support introspecting federated SDL. If this is the case, directives are stripped from the subgraph schema.
```

### Stopping an attached process

> If you stop an attached `rover dev` process (using `ctrl+c`), the supergraph will de-compose the removed subgraph and reload the router.

Reviewer: ```suggestion
If you stop an attached `rover dev` process (pressing `CTRL+C`), the supergraph will de-compose the removed subgraph and reload the router.
```
